Backports from CPython + project refresh

frenzymadness opened the pull request:
Some of the patches were applied cleanly, and some were not. The CI config is tested in my fork. I'm gonna take a look at readme.

JelleZijlstra committed:
compileall: Fix typos in docstring (GH-19810)

methane committed:
bpo-43651: Fix test_compileall with PEP 597 (GH-25128)

stefanhoelzl committed:
bpo-44666: Use default encoding as fallback for compile_file (GH-27236)

When sys.stdout.encoding is None compile_file will fall back to
sys.getdefaultencoding to encode/decode error messages.

Co-authored-by: Stefan Hoelzl <stefan.hoelzl@posteo.de>
Co-authored-by: Mickaël Schoentgen <contact@tiger-222.fr>

vishalpandeynits committed:
Fixed documentation typo in compileall.py (GH-29912)

gpshead committed:
GH-84559: Deprecate fork being the multiprocessing default. (#100618)

This starts the process. Users who don't specify their own start method
and use the default on platforms where it is 'fork' will see a
DeprecationWarning upon multiprocessing.Pool() construction or upon
multiprocessing.Process.start() or concurrent.futures.ProcessPool use.

See the related issue and documentation within this change for details.

hetmankp committed:
gh-105931: Fix surprising compileall stripdir behaviour (GH-108671)

Before, the '-s STRIPDIR' option on
compileall lead to some surprising results as it only strips away
path components that match, but leaves alone the non-matching ones
interspersed in between.

For example, with: python -m compileall -s/path/to/another/src
/path/to/build/src/file.py

The resulting written path will be: build/file.py

This fix only strips directories that are a fully matching prefix of the
source path. If a stripdir is provided that is not a valid prefix, a
warning will be displayed (which can be silenced with '-qq').

pitrou committed:
bpo-36786: Run compileall in parallel during "make install"
 (GH-13078)

paulproteus committed:
bpo-40692: Run more test_concurrent_futures tests (GH-20239)

In the case of multiprocessing.synchronize() being missing, the
test_concurrent_futures test suite now skips only the tests that
require multiprocessing.synchronize().

Validate that multiprocessing.synchronize exists as part of
_check_system_limits(), allowing ProcessPoolExecutor to raise
NotImplementedError during __init__, rather than crashing with
ImportError during __init__ when creating a lock imported from
multiprocessing.synchronize.

Use _check_system_limits() to disable tests of
ProcessPoolExecutor on systems without multiprocessing.synchronize.

Running the test suite without multiprocessing.synchronize reveals
that Lib/compileall.py crashes when it uses a ProcessPoolExecutor.
Therefore, change Lib/compileall.py to call _check_system_limits()
before creating the ProcessPoolExecutor.

Note that both Lib/compileall.py and Lib/test/test_compileall.py
were attempting to sanity-check ProcessPoolExecutor by expecting
ImportError. In multiprocessing.resource_tracker, sem_unlink() is also absent
on platforms where POSIX semaphores aren't available. Avoid using
sem_unlink() if it, too, does not exist.

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

ammaraskar committed:
bpo-34990: Treat the pyc header's mtime in compileall as an
 unsigned int (GH-19708)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Co-authored-by: Stéphane Wirtel <stephane@wirtel.be>

blueyed committed:
Fix missing space with help for `-m compileall -o` (GH-27591)

FFY00 committed:
bpo-40447: accept all path-like objects in
 compileall.compile_file (#19883)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Filipe Laíns <lains@archlinux.org>
Signed-off-by: Filipe Laíns <lains@riseup.net>
Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>
Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

frenzymadness committed:
mp_context argument for ProcessPoolExecutor works for 3.7+

frenzymadness committed:
Drop Pythons older than 3.6, add new ones

frenzymadness committed:
Drop coverage from testing

frenzymadness committed:
Migrate from Travis to Github actions

frenzymadness committed:
Update readme, drop old ToDo list

hroncok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+on:
+  push:
+    branches:
+    - "master"
+  pull_request:
+    branches:
+    - "master"
```
What is the idea behind only running the CI on the master branch?

hroncok replied:
E.g. the actions in your fork had `*` in here. So it's harder to verify they work here, because the commits are different.

frenzymadness replied:
I had "*" to test that the configuration works. I think it is quite common to have the actions configured only for the master branch and PRs targeting the master branch.

hroncok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+      uses: fedora-python/tox-github-action@main
+      with:
+        tox_env: ${{ matrix.tox_env }}
+        dnf_install: "python3-test python3.*-test pypy3-test"
```
```suggestion
        dnf_install: "python3*-test pypy3*-test"
```

frenzymadness replied:
Are you sure?
```
# dnf install python3-test python3.*-test pypy3.*-test
Installing:
 pypy3.10-test                 x86_64     7.3.13-1.3.10.fc38        updates              11 M
 pypy3.9-test                  x86_64     7.3.13-1.3.9.fc38         updates             8.3 M
 python3-test                  x86_64     3.11.8-1.fc38             updates-testing      15 M
 python3.10-test               x86_64     3.10.13-2.fc38            updates              10 M
 python3.12-test               x86_64     3.12.2-1.fc38             updates-testing      16 M
 python3.13-test               x86_64     3.13.0~a3-1.fc38          updates              16 M
Upgrading:
 python3                       x86_64     3.11.8-1.fc38             updates-testing      28 k
 python3-libs                  x86_64     3.11.8-1.fc38             updates-testing     9.6 M
…
Install  42 Packages
Upgrade   2 Packages
```
vs
```
# dnf install python3*-test pypy3*-test
Installing:
 pypy3.10-test                 x86_64   7.3.13-1.3.10.fc38            updates            11 M
 pypy3.9-test                  x86_64   7.3.13-1.3.9.fc38             updates           8.3 M
 python3-gensim-test           noarch   0.10.0-33.fc38                fedora            1.7 M
 python3-jupyter-kernel-test   noarch   0.6.0-1.fc38                  updates            31 k
 python3-samba-test            x86_64   2:4.18.10-1.fc38              updates           1.4 M
 python3-subunit-test          noarch   1.4.2-2.fc38                  fedora            104 k
 python3-test                  x86_64   3.11.8-1.fc38                 updates-testing    15 M
 python3.10-test               x86_64   3.10.13-2.fc38                updates            10 M
 python3.12-test               x86_64   3.12.2-1.fc38                 updates-testing    16 M
 python3.13-test               x86_64   3.13.0~a3-1.fc38              updates            16 M
Upgrading:
 glib2                         x86_64   2.76.6-1.fc38                 updates           2.8 M
 python3                       x86_64   3.11.8-1.fc38                 updates-testing    28 k
 python3-libs                  x86_64   3.11.8-1.fc38                 updates-testing   9.6 M
…
Install  232 Packages
Upgrade    3 Packages
```

hroncok replied:
It's not trivial to add *tested* suggestions here. Added a new one, hopefully, better. Closed this thread as resolved.

hroncok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+        dnf_install: "python3-test python3.*-test pypy3-test"
+    strategy:
+      matrix:
+        tox_env: [py36, py37, py38, py39, py310, py311, py312, py313, pypy3]
```
```suggestion
        tox_env: [py36, py37, py38, py39, py310, py311, py312, py313, pypy39, pypy310]
```

hroncok commented on tox.ini:
```diff
@@ -1,20 +1,9 @@
 [tox]
-envlist = begin,py35,py36,py37,py38,py39,py310,pypy3,end
+envlist = py36,py37,py38,py39,py310,py311,py312,py313,pypy3
```
```suggestion
envlist = py36,py37,py38,py39,py310,py311,py312,py313,pypy39,pypy310
```

hroncok reviewed (commented):
I've looked at the infra changes only.

I suggested testing with both PyPy versions we support in Fedora (but it is not necessary).

What is the motivation for * Drop coverage from testing*? Was it not working?

hroncok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+  tox_test:
+    name: Tox test
+    steps:
+    - uses: actions/checkout@v3
```
The Actions in you fork report:

> Node.js 16 actions are deprecated. Please update the following actions to use Node.js 20: actions/checkout@v3.

Can we use v4 here to avoid the warning?

frenzymadness replied:
Done.

frenzymadness commented:
> What is the motivation for * Drop coverage from testing*? Was it not working?

I believe nobody is using the results of coverage anymore. Also, to get precise numbers, we'd have to find a way how to run the tests under root which is not worth it I think.

hroncok commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+      uses: fedora-python/tox-github-action@main
+      with:
+        tox_env: ${{ matrix.tox_env }}
+        dnf_install: "python3-test python3.*-test pypy3-test"
```
```suggestion
        dnf_install: "python3-test python3.*-test pypy3*-test"
```

hroncok commented:
I know about this, but it hasn't been the most urgent thing last week. Will try to get to this soon.

frenzymadness commented:
It's not urgent and I might ask somebody else to take a look as well.

hroncok commented:
The backports commits are complete and look ok.

The 3.7 compact commit is OK as well.

I think this is ready to be fixup-0autosquashed, eyeballed and merged.

frenzymadness commented:
Rebased, squashed. The commits changing tox.ini are not perfect because I removed begin/end envs from envlist sooner than the config for these environments but I think it does not matter.

frenzymadness commented:
https://github.com/frenzymadness/compileall2/actions/runs/8373979305
